pytorch_lightning/trainer/data_loading.py

```diff
@@ -113,35 +113,44 @@ def auto_add_sampler(self, dataloader: DataLoader, train: bool) -> DataLoader:
         need_dist_sampler = (self.use_ddp or self.use_ddp2 or self.use_horovod or self.use_tpu)
 
         if self.replace_sampler_ddp and need_dist_sampler:
+            if not isinstance(dataloader.sampler, (SequentialSampler, RandomSampler)):
+                raise MisconfigurationException(
+                    'You seem to have configured a sampler in your DataLoader. This will be replaced '
+                    'by `DistributedSampler` since `replace_sampler_ddp` is True and you are using '
+                    'distributed training. Either remove the sampler from your DataLoader or set '
+                    '`replace_sampler_ddp`=False if you want to use your custom sampler.')
+
             skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
 
             dl_args = {
                 k: v for k, v in dataloader.__dict__.items() if not k.startswith('_') and k not in skip_keys
             }
 
-            if self.use_tpu:
-                sampler = DistributedSampler(
-                    dataloader.dataset,
-                    num_replicas=xm.xrt_world_size(),
-                    rank=xm.get_ordinal(),
-                )
-            elif self.use_horovod:
-                sampler = DistributedSampler(dataloader.dataset,
-                                             num_replicas=hvd.size(),
-                                             rank=hvd.rank())
-            else:
-                world_size = {
-                    'ddp': self.num_nodes * self.num_processes,
-                    'ddp2': self.num_nodes,
-                    'ddp_cpu': self.num_processes * self.num_nodes
-                }
-                sampler = DistributedSampler(
-                    dataloader.dataset,
-                    num_replicas=world_size[self.distributed_backend],
-                    rank=self.proc_rank,
-                )
-
-            dl_args['sampler'] = sampler
+            def get_distributed_sampler(dataloader):
```

Comment:
This refactoring is unrelated to PR. Done only because the build complained that the function complexity was becoming high (because of the if condition added above)

Reply:
Even though it bypasses the warning message about the complexity, it completely misses the point. 
By wrapping the code into this inner function the code gets even more complex. 
What the warning tried to tell you is move it into a separate function. 
Let's do that instead? :)

Reply:
Sure, you want me to make it a method in `TrainerDataLoadingMixin`? I'll make it a private method?
The reason I made it an inner function was because only this function seemed to be using it. But agree that what you're suggesting is more readable :)

Reply:
yes very reasonable choice in my opinion. thanks.